### PR TITLE
Fix Acorn VIDC sample playback to reduce severe aliasing

### PR DESCRIPTION
**Pull request by Lord-Nightmare**
Fix Acorn VIDC sample playback to reduce severe aliasing by correctly sequencing the 8 "channels", and properly add the FIFO and fix the MEMC-side code to fill said FIFO. [Lord Nightmare]

### CODE REVIEW

File: src/devices/machine/acorn_vidc.cpp [hunk 1]
```diff
@@ -419,15 +438,12 @@ inline void acorn_vidc10_device::refresh_stereo_image(u8 channel)
...
-	const float r_gain_settings[8] = { 1.0f, 0.0f, 0.34f, 0.66f, 1.0f, 1.34f, 1.66f, 2.0f };
+	const float l_gain_settings[8] = { 2.0f, 2.0f, 1.66f, 1.34f, 1.0f, 0.66f, 0.34f, 0.0f };
+	const float r_gain_settings[8] = { 0.0f, 0.0f, 0.34f, 0.66f, 1.0f, 1.34f, 1.66f, 2.0f };
 
 	const float l_gain = l_gain_settings[m_stereo_image[channel]] * m_sound_input_gain;
 	const float r_gain = r_gain_settings[m_stereo_image[channel]] * m_sound_input_gain;
 	LOGMASKED(LOG_STEREO, "%d: %02x -> L %f R %f\n", channel, m_stereo_image[channel], l_gain, r_gain);
-
-	m_mixer[0]->set_input_gain(channel, l_gain);
-	m_mixer[1]->set_input_gain(channel, r_gain);
```
Comment (angelosa) [inline src/devices/machine/acorn_vidc.cpp:430]: If mixer is unnecessary then it can be removed in the chain. Also is the setting 0 confirmed on real HW to be same as 1?

File: src/devices/machine/acorn_vidc.cpp [hunk 2]
```diff
@@ -450,25 +466,48 @@ void acorn_vidc10_device::sound_frequency_w(u32 data)
...
-	// assume a value of zero is invalid (ppcar POST setup)
+	// TODO: does this test bit also clear the fifo?
+	// VERIFY: assume a value of zero is invalid (ppcar POST setup)?
 	if (m_sound_mode == true && m_sound_frequency_latch)
 	{
-		// TODO: Range is between 3 and 256 usecs
-		double sndhz = get_sound_clock() / ((m_sound_frequency_latch & 0xff) + 2);
+		// Valid range is between 3(load of 0x2) and 256(load of 0xff) usecs
+		// loads of 1 and 0 are invalid and could cause undesirable behavior
+		// presumably by emptying the vidc fifo faster than the memc can
```
Comment (angelosa) [inline src/devices/machine/acorn_vidc.cpp]: Lack of uppercase capitalization went haywire here (`VIDC`, `FIFO`, `MEMC`). The note also doesn't apply to the frequency but to the `if` itself, imho I'd just add that 0 and 1 are invalid according to documentation.

Also, drop that `// VERIFY:` modification entirely. _Assume_ already implies that would need real HW test plus I've made zero `m_sound_frequency_latch` to not set any timer for a very specific reason: it crashes emulation randomly otherwise in `ppcar` (before it actually can setup the real frequency setting of 0x21). 

Also `// VERIFY:` is a non standard comment token, barely [used](https://github.com/mamedev/mame/blob/cebbc1f39e7f81ca1bd27e7f293d6d1a9ab44a20/src/mame/dec/rainbow.cpp#L2491) by stuff for no [real](https://github.com/mamedev/mame/blob/cebbc1f39e7f81ca1bd27e7f293d6d1a9ab44a20/src/mame/dataeast/sidepckt.cpp#L620) reason.

**Comment by angelosa** [comment]
There's still lots of ringing with this, which is imo unpleasant just like `master` tip. Also frequency is off.

https://github.com/user-attachments/assets/c0473e12-b2c3-4645-a52a-875025f71ca8

https://www.youtube.com/watch?v=YLAMT65e2mQ

File: src/devices/machine/acorn_vidc.cpp (hunk 1 above)
Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp:430]: Unconfirmed. The reason I changed it is because it fits a specific pattern, if you look at the way the time division muxing works. If you want, I can change it back.

File: src/devices/machine/acorn_vidc.cpp (hunk 2 above)
Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp]: I'm going to considerably clean this up, and get rid of the VERIFY

File: src/devices/machine/acorn_vidc.cpp (hunk 1 above)
Comment (angelosa) [reply src/devices/machine/acorn_vidc.cpp:430]: Yes, change it back.

Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp:430]: done.

File: src/devices/machine/acorn_vidc.cpp (hunk 2 above)
Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp]: I cleaned up the comments.

Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp]: I ended up reverting the sample rate change entirely, as it turned out it was incorrect due to another bug.

**Comment by Lord-Nightmare** [comment]
> There's still lots of ringing with this, which is imo unpleasant just like `master` tip. Also frequency is off.
> 2026-09-01.11-19-11.mp4
> 
> https://www.youtube.com/watch?v=YLAMT65e2mQ

I ended up reverting the sample rate change, since it was wrong; it appeared to be correct due to a different, now fixed bug.

File: src/devices/machine/acorn_vidc.cpp [hunk 3]
```diff
@@ -667,6 +687,26 @@ void arm_vidc20_device::device_reset()
...
+		m_sound_fifo_channel++;
+		m_sound_fifo_channel&=7;
+	}
+	else if (m_dac_serial_mode)
+	{
+		s16 sample = m_sound_fifo.dequeue()&0xff;
+		sample |= (u16)m_sound_fifo.dequeue()<<8;
+		write_dac32((m_sound_fifo_channel&2)>>1,sample);
+		m_sound_fifo_channel+=2;
+		m_sound_fifo_channel&=7;
```
Comment (angelosa) [inline src/devices/machine/acorn_vidc.cpp]: Beyond the lack of spacing, what's this and why you just don't use an XOR for driving left and right DACs?

**Comment by angelosa** [comment]
`ppcar` is broken with this, I thought that was obvious by reading the source code ...

```diff
diff --git a/src/mame/acorn/ssfindo.cpp b/src/mame/acorn/ssfindo.cpp
index b6b337b2c8c..6fe010e2103 100644
--- a/src/mame/acorn/ssfindo.cpp
+++ b/src/mame/acorn/ssfindo.cpp
@@ -648,7 +648,7 @@ void ssfindo_state::ppcar(machine_config &config)
 
        m_maincpu->set_addrmap(AS_PROGRAM, &ssfindo_state::ppcar_map);
        // sets external sclk, unverified value
-       m_vidc->set_ext_sclk(24'000'000 / (24 * 2));
+       m_vidc->set_ext_sclk(24'000'000 / 24);
 
        subdevice<qs1000_device>("qs1000")->set_external_rom(false); // ppcar has no external ROM
        subdevice<i8052_device>("qs1000:cpu")->set_disable(); // internal ROM hasn't been dumped yet
```

File: src/devices/machine/acorn_vidc.cpp (hunk 3 above)
Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp]: The serial DAC reads two bytes from the FIFO and sends those as 16 linear bits to the left channel, then reads another two bytes from the FIFO and sends those to the right channel; unlike the analog 'sdac' mode, where it sends one byte to the u-law table, then uses the stereo imaging register for that time division 'channel' to decide what percentage of that byte gets sent to left and right outputs.

Comment (angelosa) [reply src/devices/machine/acorn_vidc.cpp]: That still doesn't answer the question. Why this needs to be this convoluted, what's the point?

**Comment by Lord-Nightmare** [comment]
> `ppcar` is broken with this, I thought that was obvious by reading the source code ...
> 
> ```diff
> diff --git a/src/mame/acorn/ssfindo.cpp b/src/mame/acorn/ssfindo.cpp
> index b6b337b2c8c..6fe010e2103 100644
> --- a/src/mame/acorn/ssfindo.cpp
> +++ b/src/mame/acorn/ssfindo.cpp
> @@ -648,7 +648,7 @@ void ssfindo_state::ppcar(machine_config &config)
>  
>         m_maincpu->set_addrmap(AS_PROGRAM, &ssfindo_state::ppcar_map);
>         // sets external sclk, unverified value
> -       m_vidc->set_ext_sclk(24'000'000 / (24 * 2));
> +       m_vidc->set_ext_sclk(24'000'000 / 24);
>  
>         subdevice<qs1000_device>("qs1000")->set_external_rom(false); // ppcar has no external ROM
>         subdevice<i8052_device>("qs1000:cpu")->set_disable(); // internal ROM hasn't been dumped yet
> ```

I applied this, which fixes the issue, but I am still running into a lot of confusion (bad documentation) about the polarity of the external-SCLK-enable bit in the VIDC20, as well as exactly how many divided-down 'sample clocks' it takes to shift out one or two 16-bit linear samples serially.

File: src/devices/machine/acorn_vidc.cpp (hunk 3 above)
Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp]: This is now fixed to be much less convoluted.

File: src/devices/machine/acorn_vidc.cpp [hunk 4]
```diff
@@ -667,6 +756,26 @@ void arm_vidc20_device::device_reset()
...
+		m_sound_fifo_channel++;
+		m_sound_fifo_channel &= 7;
+	}
+	else if (m_dac_serial_mode)
+	{
+		s16 sample = m_sound_fifo.dequeue() & 0xff;
+		sample |= (u16)m_sound_fifo.dequeue() << 8;
+		write_dac32((m_sound_fifo_channel & 1), sample);
+		m_sound_fifo_channel++;
+		m_sound_fifo_channel &= 7;
```
Comment (angelosa) [inline src/devices/machine/acorn_vidc.cpp]: Again: unless the HW expects the counter to change dynamically (*) this can be expressed by just XOR-ing. We are already short in performance here, don't add extra unnecessary churn.

(*) and I doubt it does, more likely it resets the counter when switching over. Plus, Risc OS 4 changes sound system at boot time only, it doesn't switch over when issuing a `*Configure SoundSystem 8bit`

**Comment by angelosa** [comment]
Otherwise: this now sounds mergeable, I however have no domain knowledge about the filter mods.

File: src/devices/machine/acorn_vidc.cpp (hunk 4 above)
Comment (Lord-Nightmare) [reply src/devices/machine/acorn_vidc.cpp]: I left it the way it was because I was afraid some developer or demoscener would blindly change the mode register mid-fifo-pull and expect it to actually work properly; also this way the inverse of the low 2 bits m_sound_fifo_channel counter should give a pretty accurate estimate of how much is left in the FIFO...
but if you really want it to be ^= 1 I can do that, hold on.

**Comment by angelosa** [comment]
May as well test the remaining QA with this, can't be bothered in doing yet another .diff git apply for `stuntr2k` engine noise ...

File: src/devices/machine/acorn_vidc.cpp [hunk 5]
```diff
@@ -667,6 +756,25 @@ void arm_vidc20_device::device_reset()
...
+	if (m_sdac)
+	{
+		write_dac(m_sound_fifo_channel & 7, m_sound_fifo.dequeue());
+		m_sound_fifo_channel++;
+		m_sound_fifo_channel &= 7;
+	}
+	else if (m_dac_serial_mode)
+	{
+		s16 sample = m_sound_fifo.dequeue() & 0xff;
+		sample |= (u16)m_sound_fifo.dequeue() << 8;
```
Comment (angelosa) [inline src/devices/machine/acorn_vidc.cpp:771]: Why this is a thing? Why it needs the `u16` cast?

File: src/devices/machine/acorn_vidc.cpp [hunk 6]
```diff
@@ -786,13 +894,9 @@ u32 arm_vidc20_device::get_sound_clock()
...
 	{
-		return m_ext_sclk;
+		return m_ext_sclk / 24;
 	}
-
-	// 32-bit mode doubles clock rate
-	const u32 divider = 8 >> get_dac_mode();
-
-	return clock() / 24 / divider;
+	return (clock() / 24);
```
Comment (angelosa) [inline src/devices/machine/acorn_vidc.cpp:899]: Imagine testing `a7000p -bios 2 topbanan` and finding out it regressed out of a frequency bug.

File: src/devices/machine/arm_iomd.cpp [hunk 7]
```diff
@@ -866,13 +866,13 @@ void arm_iomd_device::sound_drq(int state)
...
+			m_sndcur += 4;
+		}
 
-		for (int ch = 0; ch < 2; ch++)
-			m_vidc->write_dac32(ch, (m_host_space->read_word(m_sndcur + ch * 2)));
-
-		m_sndcur += 4;
-
-		if (m_sndcur >= m_sndend)
+		if (m_sndcur > m_sndend)
```
Comment (angelosa) [inline src/devices/machine/arm_iomd.cpp:875]: This is obviously `>=` not `>`. Test case `a7000p` startup beep.
